cra-network-aware-service-worker-loading/src/sw-template.js

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+if ('function' === typeof importScripts) {
+  importScripts(
+    'https://storage.googleapis.com/workbox-cdn/releases/4.3.1/workbox-sw.js'
+  );
+  /* global workbox */
+  if (workbox) {
+    console.log('Workbox is loaded!');
+
+    /* injection point for manifest files.  */
+    workbox.precaching.precacheAndRoute([]);
+
+    /* custom cache rules*/
+    workbox.routing.registerNavigationRoute('/index.html', {
+      blacklist: [/^\/_/, /\/[^\/]+\.[^\/]+$/],
+    });
+
+    workbox.routing.registerRoute(
+      /\.(?:png|gif|jpg|jpeg)$/,
+      new workbox.strategies.CacheFirst({
+        cacheName: 'images',
+        plugins: [
+          new workbox.expiration.Plugin({
+            maxEntries: 60,
+            maxAgeSeconds: 30 * 24 * 60 * 60, // 30 Days
+          }),
+        ],
+      })
+    );
+  } else {
+    console.log('Workbox could not be loaded. No Offline support');
+  }
+}
+
+const CACHE_VERSION = 1;
+
+// Shorthand identifier mapped to specific versioned cache.
+const CURRENT_CACHES = {
+  DYNAMIC_NAME: 'ect-dynamic-v' + CACHE_VERSION
+};
+
+const matchFunction = ({url, event}) => {
+  return new RegExp('max-res|medium-res|min-res|4g-video');
+};
+
+const expirationManager = new workbox.expiration.CacheExpiration(
+  CURRENT_CACHES.DYNAMIC_NAME,
+  {
+    maxAgeSeconds: 24 * 60 * 60, // 1 Day
+    maxEntries: 50
+  }
+);
+
+const handler = async ({url, event}) => {
+  console.log('[sw handler] requesting event.request.url => ', event.request.url);
+  const cache = await caches.open(CURRENT_CACHES.DYNAMIC_NAME);
+  const response = await cache.match(event.request);
```

Comment:
I believe the goal is to treat all of the possible variations on a request URL as if they were a single, normalized cache key, and if there's a match for that normalized cache key, return the cached result. So just using `event.request` here isn't sufficient (since that might contain `max-res`, while there is a cache match for `medium-res`).

I'd recommend taking a look at https://developers.google.com/web/tools/workbox/guides/using-plugins#custom_plugins, and specifically the `cacheKeyWillBeUsed` callback. You should be able to use that to normalize the cache key for both `mode: 'read'` and `mode: 'write'`, and then just use Workbox's built-in `CacheFirst` strategy here.

Something like (untested):

```js
const expirationPlugin = new workbox.expiration.Plugin({
  maxEntries: 50,
  maxAgeSeconds: 24 * 60 * 60,
});

// See https://developers.google.com/web/tools/workbox/guides/handle-third-party-requests#force_caching_of_opaque_responses
const cacheOpaqueResponsesPlugin = new workbox.cacheableResponse.Plugin({
  statuses: [0, 200]
});

// See https://developers.google.com/web/tools/workbox/guides/using-plugins#custom_plugins
const normalizeCacheKeyPlugin = {
  cacheKeyWillBeUsed: async ({request, mode}) {
    // Normalize by removing the 'max-res', 'medium-res', or 'min-res' portion of the URL.
    // So 'https://example.com/image-max-res.jpg' and 'https://example.com/image-min-res.jpg'
    // would both end up using the same normlized cache key, 'https://example.com/image-.jpg'
    // Because we ignore `mode`, this will be done for both reads and writes.
    return request.url.replace(/max-res|medium-res|min-res/, '');
  },
};


const responsiveImagesStrategy = new workbox.strategies.CacheFirst({
  cacheName: 'responsive-images',
  plugins: [expirationPlugin, cacheOpaqueResponsesPlugin, normalizeCacheKeyPlugin],
});

workbox.routing.registerRoute(
  // See https://developers.google.com/web/tools/workbox/guides/route-requests#matching_a_route_with_a_regular_expression
  // You could use a different matching strategy if you'd prefer, but for this example, assume that we want to match
  // anything served by the Glitch CDN.
  new RegExp('^https://cdn\\.glitch\\.com/'),
  responsiveImagesStrategy
);
```

Reply:
Thank you for your information. I'll implement that way.

Reply:
I have one question.
If resources include video and image tags, I think there might be conflict between cache response type.
For example video resource is cached but image tag is avaliable then image tag will receive the cached video resource, which does not accept it.
Please give me some suggestion for it.

Reply:
I strongly question whether caching fundamentally different `Content-Type`s using the same cache key is a good approach. Do you have a strong requirement to actually do that, or could you just do only images/only videos?

If you do need to switch between images for some cases and videos for others, then I'd recommend that you don't try to cache them using the same key, and not treat them as being interchangeable from a caching perspective.

Reply:
> I strongly question whether caching fundamentally different Content-Types using the same cache key is a good approach

I'm very supportive of Jeff's question here. To the extent possible, if we can avoid caching different resource types using the same key, that would be ideal.  Images and videos can be stored under different cache keys, but when we know we're in a "low-fi" situation, we can still perform a lookup for which of a given resource type is cached or not and serve the appropriate content.

Reply:
Thank you for your clarifying.

Reply:
Please review the last commit. Here we considered the different content type by using different cache keys.